### PR TITLE
expression: fix builtin 'CharLength' for binary string input

### PR DESCRIPTION
**Pull request by spongedu**
### What problem does this PR solve? 

Currently in TiDB, builtin `CharLength`'s behavior is not consistent with MySQL when deal with binary strings.

In MySQL:
```sql
mysql> select char_length(binary("数 据 库"));
+------------------------------------+
| char_length(binary("数 据 库"))    |
+------------------------------------+
|                                 11 |
+------------------------------------+
1 row in set (0.00 sec)
```

While in TiDB:
```sql
tidb> select char_length(binary("数 据 库"));
+------------------------------------+
| char_length(binary("数 据 库"))    |
+------------------------------------+
|                                  5 |
+------------------------------------+
1 row in set (0.00 sec)

```

### What is changed and how it works?

take special care for binary strings when create `builtinSig`

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

Side effects

Related changes

### CODE REVIEW

File: expression/builtin_string.go [hunk 1]
```diff
@@ -2147,10 +2147,34 @@ func (c *charLengthFunctionClass) getFunction(ctx sessionctx.Context, args []Exp
...
+}
+
+// evalInt evals a builtinCharLengthSig for binary string type.
+// See https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_char-length
+func (b *builtinCharLengthBinarySig) evalInt(row chunk.Row) (int64, bool, error) {
+	val, isNull, err := b.args[0].EvalString(b.ctx, row)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+	return int64(len([]byte(val))), false, nil
```
Comment (coocood) [inline expression/builtin_string.go]: Just `len(val)` will do.

**Comment by coocood** [comment]
@spongedu 
Have you considered adding a field to `charLengthSig` that indicates if it is binary?
In this way, If we support more charset in the future, we may need to define a signature for each charset.

**Comment by spongedu** [comment]
At first sight I think we need only distinguish binary strings and non-binary strings. I'm not sure about this for now. I'll look into it and make some tests today or so.

File: expression/builtin_string.go (hunk 1 above)
Comment (spongedu) [reply expression/builtin_string.go]: ok

**Pull request merged**